### PR TITLE
Refactors buildmode into submodules, and adds an "atmos" buildmode

### PR DESCRIPTION
**Pull request by Crazylemon64**
Also makes the mode selection interface far less tedious/painful

atmos_modeswitch
dirswitch_buildmode

Atmos configuration is based off of partial pressure ratios of a total pressure, as well as a temperature.
By holding down "Ctrl" while selecting the area, any unsimulated turfs in the region will then "pretend" to hold the configured gas mixture to any neighboring simulated turfs.

### CODE REVIEW

**Comment by Necaladun** [comment]
I love you crazy lemon

**Comment by Spacemanspark** [comment]
Give now.

File: code/datums/click_intercept.dm [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	holder.screen += buttons
+
+/datum/click_intercept/proc/quit()
+	holder.screen -= buttons
+	holder.click_intercept = null
+	holder.show_popup_menus = 1
+	qdel(src)
+
+/datum/click_intercept/Destroy()
+	QDEL_LIST(buttons)
```
Comment (Fox-McCloud) [inline code/datums/click_intercept.dm]: should the client reference be nulled out here?

File: code/modules/buildmode/bm_mode.dm [hunk 2]
```diff
@@ -0,0 +1,44 @@
+/datum/buildmode_mode
+	var/key = "oops"
+
+	var/datum/click_intercept/buildmode/BM
+
+	var/obj/effect/buildmode_reticule/cornerA = null
+	var/obj/effect/buildmode_reticule/cornerB = null
+
+/datum/buildmode_mode/New(datum/click_intercept/buildmode/BM)
+	src.BM = BM
```
Comment (Fox-McCloud) [inline code/modules/buildmode/bm_mode.dm]: I'm not a fan of utilizing var names that *require* `src`.

File: code/modules/buildmode/bm_mode.dm [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+/datum/buildmode_mode/proc/get_button_iconstate()
+	return "buildmode_[key]"
+
+/datum/buildmode_mode/proc/show_help(mob/user)
+	CRASH("No help defined, yell at a coder")
+
+/datum/buildmode_mode/proc/change_settings(mob/user)
+	to_chat(user, "<span class='warning'>There is no configuration available for this mode</span>")
+	return
```
Comment (Fox-McCloud) [inline code/modules/buildmode/bm_mode.dm]: dangling `return`

File: code/modules/buildmode/buildmode.dm [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	close_switchstates()
+	mode = new newmode(src)
+	mode.enter_mode(src)
+	modebutton.update_icon()
+
+/datum/click_intercept/buildmode/proc/change_dir(newdir)
+	build_dir = newdir
+	close_dirswitch()
+	dirbutton.update_icon()
+	return 1
```
Comment (Fox-McCloud) [inline code/modules/buildmode/buildmode.dm]: `TRUE`

File: code/modules/buildmode/buildmode.dm [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+/datum/click_intercept/buildmode/InterceptClickOn(user,params,atom/object)
+	mode.handle_click(user, params, object)
+
+/proc/togglebuildmode(mob/M as mob in player_list)
```
Comment (Fox-McCloud) [inline code/modules/buildmode/buildmode.dm]: no `as mob`

File: code/modules/buildmode/buildmode.dm [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+	set name = "Toggle Build Mode"
+	set category = "Event"
+
+	if(M.client)
+		if(istype(M.client.click_intercept,/datum/click_intercept/buildmode))
```
Comment (Fox-McCloud) [inline code/modules/buildmode/buildmode.dm]: spacing between `,` and `/datum/click_intercept/buildmode`

File: code/modules/buildmode/buttons.dm [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+/obj/screen/buildmode/mode/Click(location, control, params)
+	var/list/pa = params2list(params)
+
+	if(pa.Find("left"))
+		bd.toggle_modeswitch()
+	else if(pa.Find("right"))
+		bd.mode.change_settings(usr)
+	update_icon()
+	return 1
```
Comment (Fox-McCloud) [inline code/modules/buildmode/buttons.dm]: `TRUE`

File: code/modules/buildmode/buttons.dm [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+	icon_state = bd.mode.get_button_iconstate()
+
+/obj/screen/buildmode/help
+	icon_state = "buildhelp"
+	screen_loc = "NORTH,WEST+1"
+	name = "Buildmode Help"
+
+/obj/screen/buildmode/help/Click()
+	bd.mode.show_help(usr)
+	return 1
```
Comment (Fox-McCloud) [inline code/modules/buildmode/buttons.dm]: `TRUE`

File: code/modules/buildmode/buttons.dm [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+	name = "Change Dir"
+
+/obj/screen/buildmode/bdir/update_icon()
+	dir = bd.build_dir
+	return
+
+/obj/screen/buildmode/bdir/Click()
+	bd.toggle_dirswitch()
+	update_icon()
+	return 1
```
Comment (Fox-McCloud) [inline code/modules/buildmode/buttons.dm]: `TRUE`

File: code/modules/buildmode/buttons.dm [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+/obj/screen/buildmode/modeswitch/New(bld, mt)
+	modetype = mt
+	icon_state = "buildmode_[initial(modetype.key)]"
+	name = initial(modetype.key)
+	return ..(bld)
+
+/obj/screen/buildmode/modeswitch/Click()
+	bd.change_mode(modetype)
+	return 1
```
Comment (Fox-McCloud) [inline code/modules/buildmode/buttons.dm]: `TRUE`

File: code/modules/buildmode/buttons.dm [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+// used to switch between dirs
+/obj/screen/buildmode/dirswitch
+	icon_state = "build"
+
+/obj/screen/buildmode/dirswitch/New(bld, dir)
+	src.dir = dir
```
Comment (Fox-McCloud) [inline code/modules/buildmode/buttons.dm]: again with the var names

File: code/modules/buildmode/buttons.dm [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+	return 1
+
+/obj/screen/buildmode/quit
+	icon_state = "buildquit"
+	screen_loc = "NORTH,WEST+3"
+	name = "Quit Buildmode"
+
+/obj/screen/buildmode/quit/Click()
+	bd.quit()
+	return 1
```
Comment (Fox-McCloud) [inline code/modules/buildmode/buttons.dm]: `TRUE`

File: code/modules/buildmode/submodes/variable_edit.dm [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+	to_chat(user, "<span class='notice'>***********************************************************</span>")
+
+// FIXME: This needs to use a standard var-editing interface instead of
+// doing its own thing here
+/datum/buildmode_mode/varedit/change_settings(mob/user)
+	var/list/locked = list("vars", "key", "ckey", "client", "firemut", "ishulk", "telekinesis", "xray", "virus", "viruses", "cuffed", "ka", "last_eaten", "urine")
+
+	varholder = input(user,"Enter variable name:" ,"Name", "name")
+	if(varholder in locked && !check_rights(R_DEBUG,0))
+		return 1
```
Comment (Fox-McCloud) [inline code/modules/buildmode/submodes/variable_edit.dm]: `TRUE`

File: code/modules/buildmode/submodes/variable_edit.dm [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+	var/thetype = input(user,"Select variable type:" ,"Type") in list("text","number","mob-reference","obj-reference","turf-reference")
+	if(!thetype) return 1
```
Comment (Fox-McCloud) [inline code/modules/buildmode/submodes/variable_edit.dm]: newline and `TRUE`

**Comment by Crazylemon64** [comment]
de-conflicted and squashed

File: code/modules/buildmode/bm_mode.dm [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+/datum/buildmode_mode/proc/exit_mode(datum/click_intercept/buildmode/BM)
+	return
+
+/datum/buildmode_mode/proc/get_button_iconstate()
+	return "buildmode_[key]"
+
+/datum/buildmode_mode/proc/show_help(mob/user)
+	CRASH("No help defined, yell at a coder")
```
Comment (tigercat2000) [inline code/modules/buildmode/bm_mode.dm]: CRASH won't give the user a message - it'll just print a debug log. Is this intentional?

Comment (Crazylemon64) [reply code/modules/buildmode/bm_mode.dm]: Oh, I didn't consider that - I'll add a message to the user

**Comment by taukausanake** [comment]
That weird moment when three other servers merge your PR before you do... 🤔

File: code/modules/buildmode/bm_mode.dm (hunk 15 above)
Comment (Crazylemon64) [reply code/modules/buildmode/bm_mode.dm]: Corrected

**Comment by variableundefined** [comment]
finally

**Pull request merged**